src/lazyapplying.jl

```diff
@@ -3,19 +3,29 @@
 abstract type ApplyStyle end
 abstract type AbstractArrayApplyStyle <: ApplyStyle end
 struct DefaultApplyStyle <: ApplyStyle end
-struct LayoutApplyStyle{Layouts<:Tuple} <: AbstractArrayApplyStyle
-    layouts::Layouts
-end
+struct DefaultArrayApplyStyle <: AbstractArrayApplyStyle end
 
-ApplyStyle(f, args...) = DefaultApplyStyle()
+ApplyStyle(f, args::Type...) = DefaultApplyStyle()
+ApplyStyle(f, args::Type{<:AbstractArray}...) = DefaultArrayApplyStyle()
 
 struct Applied{Style, F, Args<:Tuple}
-    style::Style
     f::F
     args::Args
 end
 
-Applied(f, args...) = Applied(ApplyStyle(f, args...), f, args)
+Applied{Style}(f::F, args::Args) where {Style,F,Args<:Tuple} = 
+    Applied{Style,Core.Typeof(f),Args}(f, args)
+
+_typesof() = ()
+_typesof(a, b...) = tuple(typeof(a), _typesof(b...)...)
+_typesof(a, b) = tuple(typeof(a), typeof(b))
+_typesof(a, b, c) = tuple(typeof(a), typeof(b), typeof(c))
+combine_apply_style(f, a...) = ApplyStyle(f, _typesof(a...)...)
+combine_apply_style(f, a, b) = ApplyStyle(f, typeof(a), typeof(b))
+combine_apply_style(f, a, b, c) = ApplyStyle(f, typeof(a), typeof(b), typeof(c))
```

Comment:
I wonder if we need `combine_apply_style`.  As a reference, here is how `combine_styles` is used:

```julia
@inline function broadcasted(f, args...)
    args′ = map(broadcastable, args)
    broadcasted(combine_styles(args′...), f, args′...)
end
```

In broadcasting you need to combine styles of all arguments because they have to negotiate each other to find out the best way to do `getindex`.  But in `Applied`, you don't have this problem because function "sees" the entire arguments, not just their elements.

Reply:
I'm really just using it as a way to change from calling on a value to a type. Though it hints at a possible future refinement where combined arguments are handled via `ApplyStyle(f, ::ApplyStyle, ::ApplyStyle...)`. 

I'll leave it for now as it affects whether BLAS calls are not allocating, and so it takes care to redesign it.